zeus/monitor/carbon.py

```diff
@@ -0,0 +1,448 @@
+            raise ValueError(f"Measurement window '{key}' already exists")
+        self.current_keys.add(key)
+
+        # Synchronize execution (e.g., cudaSynchronize) to freeze at the right time.
+        if sync_execution and self.zeus_monitor.gpu_indices:
+            sync_execution_fn(self.zeus_monitor.gpu_indices, sync_with=self.sync_with)
+
+        # start window
+        self.zeus_monitor.begin_window(key)
+
+        # if there were previously no active windows, start polling process
+        if len(self.current_keys) == 1:
+            self.polling_process = self.context.Process(
+                target=_polling_process,
+                args=(
+                    self.command_q,
+                    self.finished_q,
+                    self.zeus_monitor.gpu_indices,
+                    self.zeus_monitor.cpu_indices,
+                    self.carbon_intensity_provider,
+                ),
+            )
+            self.polling_process.start()
+
+        # start subwindows
+        self.command_q.put((self.Op.BEGIN, key))
+
+    def end_window(
+        self, key: str, sync_execution: bool = True
+    ) -> CarbonEmissionMeasurement:
+        """End a measurement window and return the time, energy consumption, and carbon emission.
+
+        Args:
+                        key: Name of an active measurement window.
+                        sync_execution: Whether to wait for asynchronously dispatched computations
+                                        to finish before starting the measurement window. For instance, PyTorch
+                                        and JAX will run GPU computations asynchronously, and waiting them to
+                                        finish is necessary to ensure that the measurement window captures all
+                                        and only the computations dispatched within the window.
+        """
+        # check if begin_window has been called with key before
+        if key not in self.current_keys:
+            raise ValueError(f"Measurement window '{key}' does not exist")
+
+        # Synchronize execution (e.g., cudaSynchronize) to freeze at the right time.
+        if sync_execution and self.zeus_monitor.gpu_indices:
+            sync_execution_fn(self.zeus_monitor.gpu_indices, sync_with=self.sync_with)
+
+        # end window
+        self.command_q.put((self.Op.END, key))
+
+        # continue fetching until you find the key you have received
+        while key not in self.finished_keys:
+            try:
+                (
+                    retrieved_key,
+                    gpu_carbon_emission,
+                    cpu_carbon_emission,
+                ) = self.finished_q.get_nowait()
+                self.finished_keys[retrieved_key] = (
+                    gpu_carbon_emission,
+                    cpu_carbon_emission,
+                )
+                self.current_keys.remove(key)
+            except queue.Empty:
+                pass
```

Comment:
`Or is this a design considering future multi-processing with a single polling process?`

Yes, I designed it this way because I was thinking in a situation with concurrent `end_window`s, the result you get in the end_window might be some other end_window's result. I think in that situation there is a deadlock, so I added this non-blocking way of fetching results to prevent that. At the time I was writing this, I was thinking that even though it is fetching at a high-frequency, the amount of time this loop will be running/busy-waiting is not that significant (but I may be wrong).

I think in a non-concurrent context, the implementation you have suggested is functionally equivalent and is better, and I originally had that until I thought that there might be issues once multithreading/multiprocessing is introduced.